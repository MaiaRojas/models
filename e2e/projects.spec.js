const mongoose = require('mongoose');
const { ProjectSchema } = require('schemas')(mongoose);
const Project = require('../src/models/Project')(mongoose, ProjectSchema);


describe('e2e::projects', () => {
  const projectJSON = {
    slug: 'cipher',
    repo: 'Laboratoria/curricula-js',
    path: 'projects/01-cipher',
    version: '1.0.0',
    parserVersion: '1.1.1',
    prefix: 1,
    title: 'Cifrado CÃ©sar',
    rubric: '2',
    locale: 'es-ES',
    track: 'js',
    skills: {},
  };

  beforeAll(async () => {
    await mongoose.connect(global.__MONGO_URI__, { useNewUrlParser: true });
  });

  afterAll(async () => {
    await mongoose.disconnect();
  });

  beforeEach(async () => {
    await mongoose.connection.db.dropDatabase();
    await Project.createIndexes();
  });

  it('should allow to save first version of new project', () => {
    const project = new Project(projectJSON);
    return project.save();
  });

  it('should fail to save when project/version already exists', () => {
    const project = new Project({ ...projectJSON, slug: 'cipher-bar' });
    const project2 = new Project({ ...projectJSON, slug: 'cipher-bar' });

    return project.save()
      .then(() => project2.save())
      .catch(err => expect(err.message).toMatchSnapshot());
  });

  it('should allow to save new version of existing project', () => {
    const project = new Project({ ...projectJSON, slug: 'cipher-baz' });
    const project2 = new Project({ ...projectJSON, slug: 'cipher-baz', version: '2.2.1' });

    return project.save()
      .then(() => project2.save());
  });

  it('should allow to save same version of different projects', () => {
    const project = new Project({ ...projectJSON, slug: 'cipher-qux' });
    const project2 = new Project({ ...projectJSON, slug: 'cipher-quxx' });

    return project.save()
      .then(() => project2.save());
  });

  it('should be able to query projects latest version of each project', () => {
    const project1 = new Project({ ...projectJSON, slug: 'cipher-1', version: '1.0.0' });
    const project1Version2 = new Project({ ...projectJSON, slug: 'cipher-1', version: '2.0.0' });
    const project2 = new Project({ ...projectJSON, slug: 'cipher-2', version: '1.0.0' });
    const project2Version2 = new Project({ ...projectJSON, slug: 'cipher-2', version: '2.0.0' });
    const project3 = new Project({ ...projectJSON, slug: 'cipher-3', version: '1.0.0' });

    return project1.save()
      .then(() => project1Version2.save())
      .then(() => project2.save())
      .then(() => project2Version2.save())
      .then(() => project3.save())
      .then(() => Project.findLatest())
      .then((docs) => {
        expect(docs.map(({ _id, ...doc }) => doc)).toMatchSnapshot();
      });
  });
});
