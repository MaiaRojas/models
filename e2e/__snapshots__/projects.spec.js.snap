// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e::projects should be able to query projects latest version of each project 1`] = `
Array [
  Object {
    "latestVersion": "2.0.0",
    "name": "Cifrado César",
    "slug": "cipher-1",
    "versions": Array [
      "2.0.0",
      "1.0.0",
    ],
  },
  Object {
    "latestVersion": "2.0.0",
    "name": "Cifrado César",
    "slug": "cipher-2",
    "versions": Array [
      "2.0.0",
      "1.0.0",
    ],
  },
  Object {
    "latestVersion": "1.0.0",
    "name": "Cifrado César",
    "slug": "cipher-3",
    "versions": Array [
      "1.0.0",
    ],
  },
]
`;

exports[`e2e::projects should fail to save when project/version already exists 1`] = `"E11000 duplicate key error dup key: { : \\"cipher-bar\\", : \\"1.0.0\\" }"`;
